### PR TITLE
README: revamp readme and add more comprehensive documentation

### PR DESCRIPTION
**Pull request by anakryiko**
Add explanation of concepts of retsnoop and show basic and advanced usage of it, including examples.

Signed-off-by: Andrii Nakryiko <andrii@kernel.org>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,229 +1,894 @@
...
+
+So when is LBR mode useful? There are a few typical scenarios.
+
+One of them is when you are investigating some generic error being returned
+from kernel. And such error can be returned in various different cases and it's
+not clear which one it is. This is often the case with `bpf()` syscall, where
+errors like `-EINVAL` can be returned due to dozens of various error
+conditions, and a bunch of such conditions could be checked within the same
+function. Debugging exactly which error condition is hit can maddening at
+times. LBR mode allows to get insight *into which exact if condition within the
```
Comment (Decave) [inline README.md:195]: s/LBR mode allows to get/LBR mode allows users to get

File: README.md [hunk 2]
```diff
@@ -1,229 +1,894 @@
...
+function* returned an error. See
+[third example](#peering-deep-into-functions-with-lbr) for exactly such use
+case.
```
Comment (Decave) [inline README.md:198]: not a review comment, but god I love this feature lol

File: README.md [hunk 3]
```diff
@@ -1,229 +1,894 @@
...
+
+Another common scenario is when you are trying to trace a completely unfamiliar
+part of Linux kernel code. You might know entry function, but have no clue what
```
Comment (Decave) [inline README.md:201]: s/know entry function/know the entry function

File: README.md [hunk 4]
```diff
@@ -1,229 +1,894 @@
...
+other functions it calls and where in the source code they are defined. E.g., a
+common case would be an entry function that calls into generic callback and
+it's not clear where the callback function is actually defined. In such cases
+it's very hard to know which function name globs or source code path globs to
+specify, as there could be lots of possible implementations of such callback.
```
Comment (Decave) [inline README.md:206]: s/callback/callbacks

File: README.md [hunk 5]
```diff
@@ -1,229 +1,894 @@
...
+Again, LBR can help here because it doesn't require tracing relevant functions
+to discover them.
+
+No matter what circumstances call for LBR mode, it can be activated by using
+`-R` (`--lbr`) argument. Similar to function call trace mode, LBR mode is
+independent and complimentary to default stack trace and function call trace
```
Comment (Decave) [inline README.md:212]: s/to default/to the default

File: README.md [hunk 6]
```diff
@@ -1,229 +1,894 @@
...
+modes. When the right conditions happen, `retsnoop` captures LBR data, in
+addition to stack trace and other information, and then outputs data in its own
+format. See [third example](#peering-deep-into-functions-with-lbr) for a demo
+of LBR mode.
+
+Here we'll just note that LBR mode allows customizing what kind of records CPU
+is instructed to record. It can be one of the following values: `any`,
```
Comment (Decave) [inline README.md:219]: s/records CPU is instructed to record/records the CPU is instructed to record

File: README.md [hunk 7]
```diff
@@ -1,229 +1,894 @@
...
+`any_call`, `any_return` (default), `cond, `call`, `ind_call`, `ind_jump`,
+`call_stack`, `abort_tx`, `in_tx`, `no_tx`. See Linux's
+[perf_event.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/perf_event.h#L180-L208)
+UAPI header and its `enum perf_branch_sample_type_shift`'s comments for brief
+description of each of those modes.
```
Comment (Decave) [inline README.md:224]: s/description/descriptions

File: README.md [hunk 8]
```diff
@@ -1,229 +1,894 @@
...
+
+By default, `retsnoop` assumes `any_return` LBR configuration, which records
+last N returns from functions. This is very useful for the second type of cases
```
Comment (Decave) [inline README.md:227]: s/which records last/which records the last

File: README.md [hunk 9]
```diff
@@ -1,229 +1,894 @@
...
+when we want to know which functions are being called without having to study
+code thoroughly. LBR stack will point to functions called before the deepest
```
Comment (Decave) [inline README.md:229]: s/LBR stack/The LBR stack

File: README.md [hunk 10]
```diff
@@ -1,229 +1,894 @@
...
+explicitly traced by `retsnoop` function was called, even if they weren't part
+of entry/non-entry set of functions. So you can keep iterating with `retsnoop`
+and expanding entry/non-entry function sets this way.
+
+For the first class of scenarios (complicated functions with multiple error
+conditions), LBR config `any` might be more appropriate. It records any
+function calls, returns, and jumps, both conditional and unconditional. It
+might help pinpoint exact statement *within a single big function* that is
```
Comment (Decave) [inline README.md:237]: s/exact statement/the exact statement

File: README.md [hunk 11]
```diff
@@ -1,229 +1,894 @@
...
+returning an error, as it's not bound to function boundaries.
+[LBR example](#peering-deep-into-functions-with-lbr) below is using this one to
```
Comment (Decave) [inline README.md:239]: s/LBR example/The LBR example

File: README.md [hunk 12]
```diff
@@ -1,229 +1,894 @@
...
+
+Note that LBR is a tricky business and it might not capture enough relevant
+details or sometimes it capture irrelevant details. If the latter is the case,
+you can trim it down with `--lbr-max-count` argument to emit specified number
+of most relevant entries.
+
+## Additional filters
+
+By default, `retsnoop` records any function call traces (based on entry and
+non-entry function sets) that result in triggered entry function returning
```
Comment (Decave) [inline README.md:250]: s/triggered entry function returning error/a triggered entry function returning an error

File: README.md [hunk 13]
```diff
@@ -1,229 +1,894 @@
...
+error. The "error return" is defined heuristically and is either `NULL` for
+poitner-returning functions or `-Exxx` small negative error for integer
```
Comment (Decave) [inline README.md:252]: s/poitner/pointer

File: README.md [hunk 14]
```diff
@@ -1,229 +1,894 @@
...
+returning one. This can be further adjusted, as you'll see below. But such
```
Comment (Decave) [inline README.md:253]: s/one/ones

File: README.md [hunk 15]
```diff
@@ -1,229 +1,894 @@
...
+function traces is collected globally across any process, which might be
```
Comment (Decave) [inline README.md:254]: s/is collected/are collected

File: README.md [hunk 16]
```diff
@@ -1,229 +1,894 @@
...
+inconvenient and lead to irrelevant "spam" in output.
+
+`retsnoop` allows to adjust conditions under which it captures traces, though.
```
Comment (Decave) [inline README.md:257]: s/allows to adjust/allows the user to adjust

File: README.md [hunk 17]
```diff
@@ -1,229 +1,894 @@
...
+
+### Error filters
+
+`-S` (`--success-stacks`) forces `retsnoop` to capture any function trace,
+disabling default logic of capturing only error-returning cases.
```
Comment (Decave) [inline README.md:262]: s/disabling default logic/disabling the default logic

File: README.md [hunk 18]
```diff
@@ -1,229 +1,894 @@
...
+
+`retsnoop` also allows to fine-tune which return results are considered to be
```
Comment (Decave) [inline README.md:264]: s/allows to fine-tune/allow the user to fine tune

File: README.md [hunk 19]
```diff
@@ -1,229 +1,894 @@
...
+erroneous with `-x` (`--allow-errors`) and `-X` (`--deny-errors`) arguments.
+They expect  either `Exxx`/`-Exxx` symbolic error codes (you can find all
```
Comment (Decave) [inline README.md:266]: double space after expect

File: README.md [hunk 20]
```diff
@@ -1,229 +1,894 @@
...
+recorgnized by `retsnoop` errors
```
Comment (Decave) [inline README.md:267]: s/recorgnized/recognized

In general I'd rephrase this as: (you can find all errors that are recognized by `retsnoop` [in this table](https://github.com/anakryiko/retsnoop/blob/master/src/utils.c#L10-L50)

File: README.md [hunk 21]
```diff
@@ -1,229 +1,894 @@
...
+ [in this table](https://github.com/anakryiko/retsnoop/blob/master/src/utils.c#L10-L50))
+or `NULL`.
+
+These arguments can be specified multiple times and all errors are concatenated
+into a single list.  Error denylist (`-X`) takes precedence, so if there is an
+overlap between `-x` and `-X`, `-X` wins and specified error will be ignored,
```
Comment (Decave) [inline README.md:273]: s/specified error/the specified error

File: README.md [hunk 22]
```diff
@@ -1,229 +1,894 @@
...
+those denied by `-X`, if any.
+
+### Process filtering
+
+`retsnoop` allows to narrow down a set of processes in the context of which
+data will be captured. These filters are invaluable on busy production hosts
+that have a lot of things going on at the same time, but you need to
+investigate something only within a small subset of processes.
+
+`-p` (`--pid`) and `-P` (`--no-pid`) allows to filter based on PID. Just as
```
Comment (Decave) [inline README.md:286]: s/allows to filter/allows filtering

File: README.md [hunk 23]
```diff
@@ -1,229 +1,894 @@
...
+
+`-n` (`--comm`) and `-N` (`--no-comm`) allows to control process/thread names,
+in addition or instead of PID filtering. Can be specified multiple times as
+well.
+
+### Duration filter
+
+`-L` (`--longer`) allows to specify minimal duration of triggering entry
+function execution that will be captured and reported, skipping everything that
+completed sooner. Argument should be a positive number in units of
```
Comment (Decave) [inline README.md:298]: s/Argument/This argument

File: README.md [hunk 24]
```diff
@@ -1,229 +1,894 @@
...
+Note, DWARF type information is also necessary for source code path globs to
+work.
+
+# Examples
+
+This section demonstrates various features of `retsnoop` and example output to
+get a quick feel for retsnoop's capabilities. These examples are all captured
+with help of `simfail` binary, which is part of retsnoop's repo and can be
+built from sources with `$ make simfail`. `simfail` can simulate various error
+conditions in `bpf()` syscall and is useful to get reproducible examples and
```
Comment (Decave) [inline README.md:372]: s/in `bpf()` syscall/in the `bpf()` syscall

File: README.md [hunk 25]
```diff
@@ -1,229 +1,894 @@
...
+
+## Failed per-CPU BPF array map case
+
+Let's start with `simfail`'s `bpf-bad-map-val-size-percpu-array` case, which
+attempts to create per-CPU BPF array map with value size of 1MB. This causes
+kernel to reject this because per-CPU value can't be bigger than 32KB (due to
+internal kernel per-CPU allocator limitations). But let's pretend we don't know
+about this and we are trying to figure out why it's failing.
+
+We do know that BPF map creation is done through `bpf()` syscall, which on
```
Comment (Decave) [inline README.md:383]: s/`bpf()` syscall/the `bpf()` syscall

File: README.md [hunk 26]
```diff
@@ -1,229 +1,894 @@
...
+20:19:36.372607 -> 20:19:36.372682 TID/PID 8346/8346 (simfail/simfail):
+```
+
+`20:19:36.372607 -> 20:19:36.372682` shows the timestamps of when
+`__x64_sys_bpf` (earliest activated entry function) was called and
+subsequently returned. `TID/PID 8346/8346 (simfail/simfail)` prints out thread
+and process IDs, as well as thread and process short names (so-called `comm`
+in kernel lingo, which you can get from `/proc/<pid>/comm`), respectively.
+This information lets you identify in the context of which process and thread
+stack trace was captured.
```
Comment (Decave) [inline README.md:429]: s/stack trace/the stack trace

File: README.md [hunk 27]
```diff
@@ -1,229 +1,894 @@
...
+
+After this we see stack trace with some additional information. Note that
+functions on the top are calling functions below them, forming a top-down call
+chain.
+
+Each line in a stack trace corresponds to one kernel function call. Kernel
+function name is in the center (e.g,. `__x64_sys_bpf`, `map_create`,
+`array_map_alloc`, etc), plus a hexadecimal offset (in bytes) within that
+function, which allows to pinpoint exactly assembly instruction that was
```
Comment (Decave) [inline README.md:438]: s/which allows to pinpoint exactly assembly instruction/which allows the user to pinpoint the exact assembly instruction

File: README.md [hunk 28]
```diff
@@ -1,229 +1,894 @@
...
+executed at that time. Functions that have '.' (dot) in front of them (e.g., `.
+map_create` and `.  find_and_alloc_map`) are inlined functions and they don't
+really have their own stack frame in the stack trace, but thanks to available
+DWARF debug information `retsnoop` managed to recover this information and
+present it as a `__sys_bpf` -> `map_create` -> `find_and_alloc_map` call chain,
+which is what is logically happening according to C source code.
+
+To the right of most functions we can see source code line location that was
```
Comment (Decave) [inline README.md:446]: s/we can see source code line location/we can see the source code line location

File: README.md [hunk 29]
```diff
@@ -1,229 +1,894 @@
...
+executing at the time of capturing stack trace. `kernel/bpf/syscall.c:1106:8`
```
Comment (Decave) [inline README.md:447]: s/capturing stack trace/capturing the stack trace

File: README.md [hunk 30]
```diff
@@ -1,229 +1,894 @@
...
+(next to `. map_create`) points out that C source code line #1106 in
+`kernel/bpf/syscall.c` file was the one being executed at the time of
```
Comment (Decave) [inline README.md:449]: s/at the time of/at the time of the

File: README.md [hunk 31]
```diff
@@ -1,229 +1,894 @@
...
+  1107        if (IS_ERR(map))
+  1108                return PTR_ERR(map);
+```
+
+To the left of each function that was specified through `-e` and `-a` arguments
+you'll also see two more pieces of information: function latency (e.g., 73
+microseconds for `__x64_sys_bpf`) and it's return result, which could be one of
+`-Exxx` error codes, `NULL`, `true`/`false`, some valid address like
+`0xffffc90000243000`, some other integer value, or even `void`; `retsnoop` will
+do its best to represent captured return value as meaningfully as possible,
```
Comment (Decave) [inline README.md:466]: s/represent captured return value/represent the captured return value

File: README.md [hunk 32]
```diff
@@ -1,229 +1,894 @@
...
+based on captured return value and function's type signature, if available. In
```
Comment (Decave) [inline README.md:467]: s/based on captured/based on the captured

File: README.md [hunk 33]
```diff
@@ -1,229 +1,894 @@
...
+our case, we see that innermost call to `bpf_map_alloc_percpu` returned `NULL`,
+which the got translated to `-ENOMEM` error and bubbled up all the way through
```
Comment (Decave) [inline README.md:469]: s/which the/which then

File: README.md [hunk 34]
```diff
@@ -1,229 +1,894 @@
...
+`__x64_sys_bpf` and eventually got returned to user-space as a result of
```
Comment (Decave) [inline README.md:470]: s/user-space/user space

File: README.md [hunk 35]
```diff
@@ -1,229 +1,894 @@
...
+`bpf()` syscall.
```
Comment (Decave) [inline README.md:471]: s/`bpf()` syscall/a `bpf()` syscall

File: README.md [hunk 36]
```diff
@@ -1,229 +1,894 @@
...
+
+There could be few extra markers like '!', '*', etc, at the very left side of a
```
Comment (Decave) [inline README.md:473]: s/could be few/could be a few

File: README.md [hunk 37]
```diff
@@ -1,229 +1,894 @@
...
+line, but we won't go into detail on what they mean. Usually it's pretty
+harmless to ignore them.
+
+So, in general, what did we achieve and what can we understand from
+`retsnoop`'s output? With a single invocation of `retsnoop` with few arguments
```
Comment (Decave) [inline README.md:478]: s/`retsnoop` with few arguments/`retsnoop` with a few arguments

File: README.md [hunk 38]
```diff
@@ -1,229 +1,894 @@
...
+we managed to trace `bpf()` syscall all the way down to `bpf_map_alloc_percpu`
```
Comment (Decave) [inline README.md:479]: s/trace `bpf()` syscall all the way down to/trace the `bpf()` syscall all the way down to the

File: README.md [hunk 39]
```diff
@@ -1,229 +1,894 @@
...
+function returning NULL. If we dig deeper into source code, we'll see that it
+further calls `__alloc_percpu_gfp` (which we didn't instruct `retsnoop` to
+trace, as it wasn't in either `-e` nor `-a` arguments, so we can't see that
+from the above stack trace) which got to be the one further returning `NULL`,
+indicating allocation failure. We can keep digging deeper to try to understand
```
Comment (Decave) [inline README.md:484]: s/indicating allocation failure/indicating an allocation failure

File: README.md [hunk 40]
```diff
@@ -1,229 +1,894 @@
...
+exactly why, by reading code and adjusting `retsnoop`'s set of traced function
+, but that's beyond the scope of this example.
```
Comment (Decave) [inline README.md:486]: the , at the beginning of the line should be brought up against 'function' on the prior line

File: README.md [hunk 41]
```diff
@@ -1,229 +1,894 @@
...
+
+The main point is that we can see the sequence of function calls
+leading all the way to an attempt to allocate memory for per-CPU BPF array map,
```
Comment (Decave) [inline README.md:489]: s/for per-CPU BPF/for a per-CPU BPF

File: README.md [hunk 42]
```diff
@@ -1,229 +1,894 @@
...
+and that it is exactly `bpf_map_alloc_percpu` that's failing. We also know that
+`NULL` got translated into `-ENOMEM` before bubbling up to user-space as an
```
Comment (Decave) [inline README.md:491]: s/user-space/user space

File: README.md [hunk 43]
```diff
@@ -1,229 +1,894 @@
...
+error returned from `bpf()` syscall.  Without `retsnoop` it would be a
```
Comment (Decave) [inline README.md:492]: s/from `bpf()`/from the `bpf()`

also, I suggest removing 'a' from the end of the line, and rephrasing the sentence to:


...it would be considerably harder to narrow it down so precisely...

File: README.md [hunk 44]
```diff
@@ -1,229 +1,894 @@
...
+considerably harder thing to narrow it down so precisely without the need to
+modify, recompile, and install new kernel version.
```
Comment (Decave) [inline README.md:494]: s/install new kernel version/install a new kernel version

File: README.md [hunk 45]
```diff
@@ -1,229 +1,894 @@
...
+
+Note also that we can see how did we get to our entry function
+(`__x64_sys_bpf`) through a chain of `entry_SYSCALL_64_after_hwframe` ->
+`do_syscall_64` -> `__x64_sys_bpf` calls, even though we didn't instruct
+`retsnoop` to trace first two functions. This ability to see how we got to our
```
Comment (Decave) [inline README.md:499]: s/trace first two/trace the first two

File: README.md [hunk 46]
```diff
@@ -1,229 +1,894 @@
...
+entry function without explicitly specifying possible functions with `-e` or
+`-a` is extremely useful, especially when dealing with unfamiliar parts of the
+kernel.
+
+## Tracing BPF verification flow
+
+`retsnoop` was initially designed to debug cases when kernel returns some
+generic error (like `-EINVAL`) from somewhere deep inside complex syscall (like
```
Comment (Decave) [inline README.md:507]: s/inside complex syscall/inside a complex syscall

File: README.md [hunk 47]
```diff
@@ -1,229 +1,894 @@
...
+`bpf()` syscall) and it's hard to guess that that error is. So by default
```
Comment (Decave) [inline README.md:508]: s/guess that that/guess what that

File: README.md [hunk 48]
```diff
@@ -1,229 +1,894 @@
...
+`retsnoop` will only catch call stack that end up returning error (`-Exxx` or
```
Comment (Decave) [inline README.md:509]: s/catch call stack that end up returning error/catch the call stack that ends up returning an error

File: README.md [hunk 49]
```diff
@@ -1,229 +1,894 @@
...
+`NULL`), and will try to get the deepest call stack possible to help narrow
+down the origin of the error. Just like we saw in the previous example.
+
+But `retsnoop` can do much more than that. It can trace any call stack, not
+just error-returning one. It also has a function call tracing mode, in which we
```
Comment (Decave) [inline README.md:514]: s/just error-returning one/just an error-returning one

File: README.md [hunk 50]
```diff
@@ -1,229 +1,894 @@
...
+can see how kernel control flow evolves. In this example we'll try to get an
+overview of a pretty complex workflow: successful validation of BPF program by
```
Comment (Decave) [inline README.md:516]: s/successful validation of BPF program by BPF verifier/successful validation of a BPF program by the BPF verifier

File: README.md [hunk 51]
```diff
@@ -1,229 +1,894 @@
...
-                                               . preempt_count                           (arch/x86/include/asm/preempt.h:27:9)
-           22us [NULL]        ffffffff811e4815 __alloc_percpu_gfp+0x5                    (mm/percpu.c:1894:9)
-                             ~ffffffffa1ca4065 bpf_trampoline_6442477050_1+0x65
-            3us [NULL]        ffffffff811e4035 pcpu_alloc+0x5                            (mm/percpu.c:1679:1)
-                             ~ffffffffa1ea6096 bpf_trampoline_6442477052_1+0x96
-                             ~ffffffffa0ba2513 bpf_prog_22cf752688d52e2b_fexit4+0x463
-                             ~ffffffff81155491 bpf_get_stack_raw_tp+0x51
-                             ~ffffffffa0ba2513 bpf_prog_22cf752688d52e2b_fexit4+0x463
 ```
+# after retsnoop loads, run `sudo ./simfail bpf-simple-obj` in separate terminal
```
Comment (Decave) [inline README.md:523]: s/in separate terminal/in a separate terminal

File: README.md [hunk 52]
```diff
@@ -1,229 +1,894 @@
...
+                  -d 'bpf_lsm*' -d 'push_insn' -d '__mark_reg_unknown' \
+                  -S -T
 ```
 
-If kernel image (vmlinux) is available and contains DWARF information, it's
-possible to select a set of functions using its source code location relative
-to the root. The syntax is `':<path-glob>'`. The following example will select
-all attachable functions in BTRFS-related files. Such syntax works for entry,
-allow, and deny function specifiers.
+Instead of using `bpf()` syscall as an entry point, we ask `retsnoop` to only
```
Comment (Decave) [inline README.md:598]: s/using `bpf()`/using the `bpf()`

File: README.md [hunk 53]
```diff
@@ -1,229 +1,894 @@
...
+activate tracing when `bpf_prog_load` is called. This is an entry point for BPF
```
Comment (Decave) [inline README.md:599]: s/entry point for BPF/entry point for the BPF

File: README.md [hunk 54]
```diff
@@ -1,229 +1,894 @@
...
+program validation command inside `bpf()` syscall. This allows us to ignore
```
Comment (Decave) [inline README.md:600]: s/inside `bpf()`/inside the `bpf()`

File: README.md [hunk 55]
```diff
@@ -1,229 +1,894 @@
...
+other irrelevant `bpf()` syscall invocations.
+
+Next, we ask `retsnoop` to trace all discovered (through DWARF debug info)
+functions in `kernel/bpf/\*.c` files. As we don't really know what we are
+looking for, it makes sense to "cast the net wide". Note also that I filtered
+out (with the use of `-d` arguments) few naming patterns ('bpf_lsm\*',
```
Comment (Decave) [inline README.md:606]: s/few naming patterns/a few naming patterns

File: README.md [hunk 56]
```diff
@@ -1,229 +1,894 @@
...
+ignore such case. Adding `-S` disables `retsnoop`'s default error-focused
+capturing logic.
+
+`-T` instructs `retsnoop` to collect a detailed record of all traced function
+calls and emit it in a separate table, like the one you can see above.
+
+We'll look at what that function call trace table shows in a second. But first,
+we do also see a standard stack trace at the very bottom of the output. In this
+case, it ends up with `bpf_map_kptr_off_contains` that returns `NULL`. This is
+`retsnoop` trying to be helpful to detect deepest error-returning function call
```
Comment (Decave) [inline README.md:627]: s/to detect deepest/to detect the deepest

File: README.md [hunk 57]
```diff
@@ -1,229 +1,894 @@
...
+chain (and `NULL` is considered an error by default). We don't care much about
+call stack in this example, so just ignore it in this case. Stack trace is
```
Comment (Decave) [inline README.md:629]: s/call stack/the call stack

File: README.md [hunk 58]
```diff
@@ -1,229 +1,894 @@
...
+always emitted by `retsnoop`, there is no way to suppress it. This might be
+possible in the future.
+
+Back to function call trace table. Let's shorten it some more to explain main
```
Comment (Decave) [inline README.md:633]: s/explain main elements/explain the main elements

File: README.md [hunk 59]
```diff
@@ -1,229 +1,894 @@
...
+    [...]
+    → bpf_prog_kallsyms_add
+        ↔ bpf_ksym_add                            [void]                  2.046us
+    ← bpf_prog_kallsyms_add                       [void]                  6.104us
+← bpf_prog_load                                   [5]                   374.697us
 ```
 
-## Use cases
+In the left column we see function names with one of `→`, `←`, or `↔` markers
+on the left. `→` marks entry into traced function, `←` marks and exit from a
```
Comment (Decave) [inline README.md:651]: s/into traced function/into the traced function

also,

s/marks and exit/marks an exit

File: README.md [hunk 60]
```diff
@@ -1,229 +1,894 @@
...
+function. If a function didn't have any further calls to any other traced
+function and thus is a leaf function (as far as a specified set of functions of
+interest goes; note, that in reality there could have been other functions
+called further, but they were not traced and thus captured by `retsnoop` due to
+them being omitted from `-e`/`-d` specifications). Such leaf functions, instead
+of taking two rows with separate `→` (entry) and `←` (exit) records, are
+collapsed into a single entry-exit row marked with `↔`. As calls become more
+nested, `retsnoop` indents each level of nestedness with four spaces.
+
+For each row corresponding to function exit event (so, `→`- and `↔`-marked
```
Comment (Decave) [inline README.md:661]: s/to function exit event/to a function exit event

File: README.md [hunk 61]
```diff
@@ -1,229 +1,894 @@
...
+rows), retsnoop also reports return result, similarly to default stack trace
```
Comment (Decave) [inline README.md:662]: s/reports return result/reports the return result

also

s/similarly to default/similarly to the default

File: README.md [hunk 62]
```diff
@@ -1,229 +1,894 @@
...
+view. And also similarly to stack trace view, we get a duration of a complete
```
Comment (Decave) [inline README.md:663]: s/similarly to stack/similarly to the stack

File: README.md [hunk 63]
```diff
@@ -1,229 +1,894 @@
...
+then called `bpf_prog_alloc` which further called `bpf_prog_alloc_no_stats`.
+`bpf_prog_alloc_no_stats` returned some valid kernel address
+(`0xffffc9000031e000`), which was subsequently returned from `bpf_prog_alloc`
+back to `bpf_prog_load`. We then skip a lot of function invocations for
+brevity, but eventually we get to `bpf_prog_kallsyms_add`, which calls
+`bpf_ksym_add` once. Then `bpf_ksym_add` and, subsequently,
+`bpf_prog_kallsyms_add` return with no results (they are `void` functions,
+according to BTF type information). After `bpf_prog_kallsyms_add` returned,
+`bpf_prog_load` completes and returns a successful result (FD 5 in this case),
+entire invocation of `bpf_prog_load` takes almost 375 microseconds.
```
Comment (Decave) [inline README.md:676]: suggestion: I'd make this last line a separate sentence, starting with "The entire invocation..."

File: README.md [hunk 64]
```diff
@@ -1,229 +1,894 @@
...
+function calls. With `retsnoop`'s flexible mechanism to specify just the right
+subset of functions to be traced, such function traces can omit a lot of
+uninteresting function calls, if you know what you are most interested in.
+
+This tracing mode is really great to discover how kernel works and I hope
+you'll find it valuable!
+
+## Peering deep into functions with LBR
+
+To demonstrate LBR mode and how it helps in practice we'll utilize
```
Comment (Decave) [inline README.md:689]: s/utilize/utilize the

File: README.md [hunk 65]
```diff
@@ -1,229 +1,894 @@
...
+`bpf-bad-map-max-entries-array` test case from `simfail`. In this test case
+`simfail` is attempting to create BPF array map with invalid parameters,
```
Comment (Decave) [inline README.md:691]: s/to create/to create a

File: README.md [hunk 66]
```diff
@@ -1,229 +1,894 @@
...
+leading to `bpf()` syscall returning a generic `-EINVAL` error. Let's see how
```
Comment (Decave) [inline README.md:692]: s/leading to/leading to the

File: README.md [hunk 67]
```diff
@@ -1,229 +1,894 @@
...
+we can quickly narrow this down to a specific check without knowing much about
+BPF subsystem source code.
```
Comment (Decave) [inline README.md:694]: s/BPF subsystem/the BPF subsystem

File: README.md [hunk 68]
```diff
@@ -1,229 +1,894 @@
...
+                    do_syscall_64+0x2d                   (arch/x86/entry/common.c:46:12)
+    24us [-EINVAL]  __x64_sys_bpf+0x1c                   (kernel/bpf/syscall.c:4749:1)
+!   10us [-EINVAL]  __sys_bpf
 ```
 
-Either use case or at least one entry glob that matches at least one function
-has to be specified. Entry functions (those that match entry globs) are
-automatically enlisted as allowed functions.
+Looking at stack trace output we do indeed see that `__x64_sys_bpf` returned
+`-EINVAL`, so this must be what we are trying ot catch. But beyond that it
```
Comment (Decave) [inline README.md:726]: s/ot catch/to catch

File: README.md [hunk 69]
```diff
@@ -1,229 +1,894 @@
...
+CPU performed a call, return, or jump and correspondingly instruction address
+to where such call/return/jump was performed. When post-processed to turn such
+addresses into matched function and source code location (as we do for stack
+traces as well), we'll get a collection of pairs of from/to functions (and, if
+DWARF data is available, their source code locations, down to exact line within
+a specific source code file).
+
+This is what we basically see in the output above. To the left of `->` we see
+where the jump was performed from, and to the right is a destination of such
+jump. Each individual LBR record also is numbered and in the output above we
```
Comment (Decave) [inline README.md:741]: s/LBR record also is numbered/LBR record is also numbered

File: README.md [hunk 70]
```diff
@@ -1,229 +1,894 @@
...
+see `[#03]` through `[#07]`. In full output I get records numbered all the way
+to `[#31]`, because my CPU supports capturing 32 LBR records. The reason we
+don't see `[#00]` through `[#02]` is due to `retsnoop` filtering out few
```
Comment (Decave) [inline README.md:744]: s/filtering out few/filtering out the few

File: README.md [hunk 71]
```diff
@@ -1,229 +1,894 @@
...
+irrelevant LBR recordsd that are "wasted" on kernel-internal functionality
```
Comment (Decave) [inline README.md:745]: s/recordsd/records

File: README.md [hunk 72]
```diff
@@ -1,229 +1,894 @@
...
+necessary to capture LBR at all. This is unavoidable price to be able to
+capture LBR at any point with BPF: we have to waste few records (and sometimes
```
Comment (Decave) [inline README.md:747]: s/waste few/waste a few

File: README.md [hunk 73]
```diff
@@ -1,229 +1,894 @@
...
+```
+[#04] array_map_alloc_check+0x74 (kernel/bpf/arraymap.c:79:1)  ->  __sys_bpf+0xca0      (kernel/bpf/syscall.c:4629:9)
+                                                                   . map_create         (kernel/bpf/syscall.c:863:8)
+                                                                   . find_and_alloc_map (kernel/bpf/syscall.c:123:6)
+```
 
-## Filtering by process ID (PID)
+This is actually the case of having inlined function calls. Both `map_create`
+is inlined inside `__sys_bpf`, and `find_and_alloc_map` is inlined within
+`map_create`, so LBR record captured a jump to what used to be code belonging
```
Comment (Decave) [inline README.md:763]: s/so LBR record/so the LBR record

File: README.md [hunk 74]
```diff
@@ -1,229 +1,894 @@
...
+to `find_and_alloc_map` and `retsnoop` helpfully points out that it's logically
+a sequence of `__sys_bpf` -> `map_create` -> `find_and_alloc_map` calls.
+
+Apart from that, each side of an `->` arrow is a simplified version of default
+stack trace output format, except there is no return result or duration there.
+Just function name (with hexadecimal offset) and corresponding source code
+location.
+
+Ok, back to the task at hands. It's best to "unwind" LBR records from the most
```
Comment (Decave) [inline README.md:772]: s/task at hands/task at hand

File: README.md [hunk 75]
```diff
@@ -1,229 +1,894 @@
...
+recent (bottom ones, with smalles numbers) to older (top ones, with higher
+numbers).
+
+Record #3 isn't all that useful, it just shows that we exited `__sys_bpf` and
+jumped into kernel's kretprobe handler (which is what `retsnoop` is using to
```
Comment (Decave) [inline README.md:777]: s/into kernel's/into the kernel's

**Review by Decave — COMMENTED** [review]
Awesome write up. This is great stuff. Left a bunch of comments that are mostly grammar nits.

File: README.md (hunk 2 above)
Comment (anakryiko) [reply README.md:198]: I know, me too! :) At some point I almost gave up on this, but Song managed to somehow allow this ad-hoc LBR capture from BPF.

File: README.md [hunk 76]
```diff
@@ -1,229 +1,894 @@
...
+output to understand what's going on under the cover. `retsnoop` also has more
+verbose levels (`-vv` and `-vvv`), but they most probably only are useful for
+developers of `retsnoop` and for debugging.
+
+Default verbose output is extremely useful in combination with dry-run mode,
+activated with `--dry-run` argument. In this mode `retsnoop` will report all
+the actions it's going to perform (including listing which functions it
+discovered and is going to trace), but stops short of actually activating any
+of that. This way there is not even a chance to disturb production workload and
+it's a safe way to pre-check everything upfront.
```
[the remark this answers is not in the document]
Comment (anakryiko) [reply README.md:321]: intentional, but I'll see if I can just cross-reference them seamlessly

File: README.md (hunk 34 above)
Comment (anakryiko) [reply README.md:470]: indeed, wikipedia also uses "user space", fixed everywhere

**Review by anakryiko — COMMENTED** [review]
@Decave thanks a lot for a very thorough review, I apprecaite all the time and effort you spent on this! I'll rework the intro section stealing according to your suggestions! Thanks again!

File: README.md [hunk 77]
```diff
@@ -1,229 +1,894 @@
...
+
+As mentioned above, by default `retsnoop` is capturing a stack trace, with the
+deepest nested function calls that satisfy conditions (i.e., if no custom error
+filtering is specified, `retsnoop` will try to capture stack trace of a deepest
+function call chain resulting in error return). See
+[first example](#failed-per-cpu-bpf-array-map-case) for and example output in
+this mode, and how to interpret it.
+
+This mode is currently always on and can't be disabled. Other modes (function
+calls trace and LBR, described below) are complementary to default stack trace
```
[the remark this answers is not in the document]
Comment (Decave) [reply README.md:154]: I believe "function call traces" can actually mean either of the two possible definitions you described. "Function call trace" seems like the most appropriate way to phrase this though.
